components/resource-creation/ResourcePanel.tsx

```diff
@@ -1,16 +1,49 @@
-import { Center, Divider, Text } from '@mantine/core';
-import { Suspense } from 'react';
-import { useRecoilValue } from 'recoil';
+import { Button, Center, Divider, Text } from '@mantine/core';
+import { v4 as uuidv4 } from 'uuid';
+import { IconAlertCircle, IconCodePlus } from '@tabler/icons';
+import { Suspense, useState } from 'react';
+import { useRecoilState, useRecoilValue } from 'recoil';
 import { Loader } from 'tabler-icons-react';
 import { patientTestCaseState } from '../../state/atoms/patientTestCase';
 import { selectedPatientState } from '../../state/atoms/selectedPatient';
 import { getPatientNameString } from '../../util/fhir';
+import produce from 'immer';
+import CodeEditorModal from '../modals/CodeEditorModal';
 import ResourceDisplay from './ResourceDisplay';
 import ResourceSelection from './ResourceSelection';
+import { showNotification } from '@mantine/notifications';
 
 export default function ResourcePanel() {
   const selectedPatient = useRecoilValue(selectedPatientState);
-  const currentPatients = useRecoilValue(patientTestCaseState);
+  const [currentPatients, setCurrentPatients] = useRecoilState(patientTestCaseState);
+  const [isNewResourceModalOpen, setIsNewResourceModalOpen] = useState(false);
+
+  const createNewResource = (val: string) => {
+    // TODO: Validate the incoming JSON as FHIR
+    const newResource = JSON.parse(val.trim());
+    // Create a new state object using immer without needing to shallow clone the entire previous object
+    if (selectedPatient) {
+      if (!newResource.id) {
+        newResource.id = uuidv4();
+      }
+      const resourceIndex = currentPatients[selectedPatient].resources.findIndex(r => r.id === newResource.id);
+      if (resourceIndex >= 0) {
+        showNotification({
+          id: 'failed-upload',
+          icon: <IconAlertCircle />,
+          title: 'Resource Creation Failed',
+          message: `Resource ID ${newResource.id} is used for an existing resource for this patient.`,
+          color: 'red'
+        });
```

Comment:
This was a really good call! I don't know if this is in your scope of the task, but I noticed through this that if you add a resource (not a custom one) and then change its ID then it replaces the resource that had that ID with the new one...maybe we should add this notification to that case too!

Reply:
That's a good point. Will there ever be a case where we have a resource with, for example, id `123` and then we edit a different resource to have id `123` but we actually want the second resource to replace the first one (in terms of references, etc.)? If not, then I think we can easily add the notification for this case as well. @mgramigna thoughts?

Reply:
in general I don't think users will ever really be editing IDs. I'm fine with restricting behavior such that IDs need to be unique during any creation of a resource in a modal (no overwriting. do that through the edit functionality if you want to replace a resource)

you _can_ make that change here to show an error in the case Elsa mentioned too but you don't have to

Reply:
After looking at the code more, I think I will hold off on making the change here since creating a new resource and updating an existing resource both call `updateResource` (but follow different branches within the function depending on whether the id already exists for the current patient). I think making this change would require some larger changes to determine whether we are creating a new resource with an existing id because that would mean we would need to pass in information aside from the id saying that this is a new resource (not sure if that makes sense sorry - I'm looking at line 75 in ResourceDisplay.tsx where we determine whether to update/create based on whether the id is already present)

Reply:
After looking at the code more, I think I will hold off on making the change here since creating a new resource and updating an existing resource both call updateResource (but follow different branches within the function depending on whether the id already exists for the current patient). I think making this change would require some larger changes to determine whether we are creating a new resource with an existing id because that would mean we would need to pass in information aside from the id saying that this is a new resource (not sure if that makes sense sorry - I'm looking at line 75 in ResourceDisplay.tsx where we determine whether to update/create based on whether the id is already present)